jcli/src/jcli_app/vote/tally/decrypt_shares.rs

```diff
@@ -31,9 +30,54 @@ pub struct TallyDecryptWithAllShares {
     output_format: OutputFormat,
 }
 
+#[derive(StructOpt)]
+#[structopt(rename_all = "kebab-case")]
+pub struct TallyVotePlanWithAllShares {
+    /// The path to json-encoded vote plan to decrypt. If this parameter is not
+    /// specified, the vote plan will be read from the standard
+    /// input.
+    #[structopt(long = "vote-plan")]
+    vote_plan: Option<PathBuf>,
+    /// The id of the vote plan to decrypt.
+    /// Can be left unspecified if there is only one vote plan in the input
+    #[structopt(long = "vote-plan-id")]
+    vote_plan_id: Option<String>,
+    /// The minimum number of shares needed for decryption
+    #[structopt(long = "threshold", default_value = "3")]
+    threshold: usize,
+    /// The path to json-encoded necessary base64 shares. If this parameter is not
+    /// specified, the shares will be read from the standard input.
+    #[structopt(long = "shares")]
+    shares: Option<PathBuf>,
+    #[structopt(flatten)]
+    output_format: OutputFormat,
+}
+
 #[derive(Serialize)]
 struct Output {
-    result: Vec<Option<u64>>,
+    result: Vec<u64>,
+}
+
+fn read_shares_from_file<P: AsRef<Path>>(
+    share_path: &Option<P>,
+    threshold: usize,
+    proposals: usize,
+) -> Result<Vec<Vec<chain_vote::TallyDecryptShare>>, Error> {
+    let shares: Vec<Vec<String>> = serde_json::from_reader(io::open_file_read(share_path)?)?;
+    if shares[0].len() < threshold || shares.len() != proposals {
```

Comment:
There is no guarantee that all underlying vectors are of correct length if `shares[0].len() >= threshold` is satisfied. IMO it is better to move that check to the `.map()` call below. I guess this pattern might be repeated in a few other places.